### DISCUSSION TITLE
BelongsToManyMultiSelect dosn't support Policies on target entity

### DISCUSSION THREAD

**Discussion by simonbuehler** [filamentphp/filament#2050] [Help] [answered]
### Package

filament/forms

### Package Version

v2.10.39

### Laravel Version

^8.75

### Livewire Version

^2.5

### PHP Version

8

### Bug description

hi,

i have a user model with many roles ( spatie-permissions). on the edit page of the user the roles are edited using a `BelongsToManyMultiSelect` which works fine with a default or no policy.

as soon as i change the policy to use a secon injected user target object it breaks:

UserPolicy
``` php 
    /**
     * Determine whether the user can update the model.
     *
     * @param  \App\Models\User  $user
     * @param  \App\Models\User  $target 
     * @return \Illuminate\Auth\Access\Response|bool
     */
    public function update(User $user, User $target)
    {
        return !$target->hasRole('super_admin'); // do no mess with the boss ... but breaks the select
       // return $user->can('update_user'); // works fine
    }
```

the error shown:

```
SQLSTATE[HY093]: Invalid parameter number (SQL: select `name`, `id` from `roles` where `id` in (4, Bewerber)) 
```

any hint how to debug?

### Steps to reproduce

add spatie-permissions, add UserResource in Filament, Set UserPolicy to inject the target, watch it crash

### Relevant log output

```shell
fingerprint
    { "id": "4VIodEYE5OTk2KsGfNFj", "name": "filament.resources.user-resource.pages.edit-user", "locale": "de", "path": "admin/users/3/edit", "method": "GET", "v": "acj" }
serverMemo
    { "children": [], "errors": [], "htmlHash": "bf7b2e09", "data": { "record": [], "data": { "id": 3, "vorname": "asdsad", "nachname": "asdasd", "email": "xx@xx.xx", "email_verified_at": null, "current_team_id": null, "profile_photo_path": null, "created_at": "2022-03-22T10:00:35.000000Z", "updated_at": "2022-03-22T10:00:35.000000Z", "profile_photo_url": "https://ui-avatars.com/api/?name=A+a&color=7F9CF5&background=EBF4FF", "roles": [ { "id": 4, "name": "role1", "guard_name": "web", "created_at": "2022-03-21T10:49:58.000000Z", "updated_at": "2022-03-21T10:49:58.000000Z", "pivot": { "model_id": 3, "role_id": 4, "model_type": "App\\Models\\User" } }, { "id": 5, "name": "Unbestätigt", "guard_name": "web", "created_at": "2022-03-21T10:49:58.000000Z", "updated_at": "2022-03-21T10:49:58.000000Z", "pivot": { "model_id": 3, "role_id": 5, "model_type": "App\\Models\\User" } } ] }, "mountedAction": null, "mountedActionData": [], "componentFileAttachments": [], "activeRelationManager": null }, "dataMeta": { "models": { "record": { "class": "App\\Models\\User", "id": 3, "relations": [ "roles" ], "connection": "mysql" } } }, "checksum": "2baad197b3983de92990091b92a69acd8ce34e76e4d9b89f00279169c86b080d" }
updates
    [ { "type": "callMethod", "payload": { "id": "8v8x", "method": "getMultiSelectOptionLabels", "params": [ "data.roles" ] } } ]
```

**Comment by henninghorn** [comment]
Hi @simonbuehler 

I'm not sure how the usage of the BelongsToManyMultiSelect field is related to the issue.

If the user policy's `update` method returns false, then you shouldn't see a edit link in the user resource's list view.
And furthermore you shouldn't be able to open the edit page for a user.

It sounds like some sort of issue with the hasRole call instead, like the SQL error indicates.

**Comment by simonbuehler** [comment]
sorry for beeing unclear here , the icons ( edit , view ) do show or hide correct in the list view.

![grafik](https://user-images.githubusercontent.com/78061/160790661-63d4ac79-3da7-40d0-aeca-d7642c058eb6.png)

_editing_ a allowed user ( in the example any user not having the super_admin role) breaks , when removing the
```php
BelongsToManyMultiSelect::make('roles')->relationship('roles', 'name'),
```
it works

**Comment by henninghorn** [comment]
Perhaps it is a combination of that field and another field in the form which causes the issue?

Try only having the BelongsToManyMultiSelect field in the form, and see if it still breaks.

**Comment by simonbuehler** [comment]
that doesn't help, its still the livewire call to `getMultiSelectOptionLabels` with  params `[ "data.roles" ]` that fails

**Comment by danharrin** [comment]
Can I see a screen recording where you show it doesnt work, then go to your code editor, comment out the policy line, and then switch back and show it does work

BelongsToManyMultiSelect doesnt read policies at all

**Reply by simonbuehler** [reply]
here we go..

https://user-images.githubusercontent.com/78061/160816421-b673c084-c389-437b-8fd1-978899b702be.mp4

.

**Reply by danharrin** [reply]
Can you share the Flare exception link please?

**Comment by simonbuehler** [comment]
https://flareapp.io/share/W7zgEYV5#F75

**Reply by danharrin** [reply]
From the error, it looks like one of the roles has a non-numeric ID? Where does that come from?

**Reply by danharrin** [reply]
"Bewerber"

**Comment by simonbuehler** [comment]
User Model with spatie `HasRole` trait
``` php
 /**
     * A model may have multiple roles.
     */
    public function roles(): BelongsToMany
    {
        $relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            PermissionRegistrar::$pivotRole
        );

```

table roles
```
"id","name","guard_name","created_at","updated_at"
"1","super_admin","web","2022-03-21 11:43:10","2022-03-21 11:43:10"
"2","Admin","web","2022-03-21 11:49:58","2022-03-21 11:49:58"
"3","Benutzer","web","2022-03-21 11:49:58","2022-03-21 11:49:58"
"4","Bewerber","web","2022-03-21 11:49:58","2022-03-21 11:49:58"
"5","Unbestätigt","web","2022-03-21 11:49:58","2022-03-21 11:49:58"
```

table model_has_roles
```
"role_id","model_type","model_id"
"1","App\Models\User","1"
"3","App\Models\User","1"
"2","App\Models\User","2"
"4","App\Models\User","3"
"5","App\Models\User","3"
```

**Comment by simonbuehler** [comment]
maybe the *morphToMany* is making troubles?

**Comment by ChrisHardie** [comment]
Just noting that I hit this today as well. In my User policy, when I change the `update()` method to include a `hasRole()` call on the target user model being updated, trying to load an "edit user" form that contains a multi-select for roles (powered by https://github.com/3x1io/filament-user and https://github.com/bezhanSalleh/filament-shield) and the call to a Livewire endpoint to fetch the list of roles ends up throwing an `Illuminate\Database\QueryException` with the message `SQLSTATE[HY093]: Invalid parameter number (SQL: select `name`, `id` from `roles` where `id` in (3))`. Same behavior in the screen recording above where the role "X" icons appear right before the error is hit. 

As soon as I remove the policy logic, the form and the select work again.

**Comment by simonbuehler** [comment]
@danharrin any hint how to fix this?

**Reply by danharrin** [reply]
This is a complex issue and I have no clue what could be causing the error, so I'll need a repo that can reproduce the issue. Please include all necessary models, migrations & a seeder for fake data.

**Reply by simonbuehler** [reply]
https://github.com/simonbuehler/bug_demo

**Comment by danharrin** [comment, marked as answer]
Fixed by #2519, the issue is because the field is named identically to the relationship, meaning that the value was being loaded from the relationship as a collection instead of an array of IDs.

The reason why it was loaded in the first place was because the policy was checking the user role, which meant the `roles` relationship was loaded.

**Reply by simonbuehler** [reply]
Thanks, amazing work!
